clients/src/test/java/org/apache/kafka/common/network/EchoServer.java

```diff
@@ -71,49 +72,54 @@ public void renegotiate() {
     @Override
     public void run() {
         try {
-            while (true) {
+            while (!closing) {
                 final Socket socket = serverSocket.accept();
-                sockets.add(socket);
-                Thread thread = new Thread() {
-                    @Override
-                    public void run() {
-                        try {
-                            DataInputStream input = new DataInputStream(socket.getInputStream());
-                            DataOutputStream output = new DataOutputStream(socket.getOutputStream());
-                            while (socket.isConnected() && !socket.isClosed()) {
-                                int size = input.readInt();
-                                if (renegotiate.get()) {
-                                    renegotiate.set(false);
-                                    ((SSLSocket) socket).startHandshake();
-                                }
-                                byte[] bytes = new byte[size];
-                                input.readFully(bytes);
-                                output.writeInt(size);
-                                output.write(bytes);
-                                output.flush();
-                            }
-                        } catch (IOException e) {
-                            // ignore
-                        } finally {
+                synchronized (sockets) {
```

Comment:
`closing` flag should be checked inside the synchronized block. Otherwise, this block could be executed after  `closeConnections` on a socket that was already accepted, but not added to the list (this is the cause of the transient test failure). The `!closing` check in `while` is not really needed since `accept` would just throw IOException and return, but you want to keep it, it should be synchronized or made volatile.

Reply:
Once closing flag becomes true, the while loop in EchoServer#run() should exit.
What about the last assertion in testServerDisconnect() ?
```
        assertEquals("hello", blockingRequest(node, "hello"));
```
It seems the above should be dropped - otherwise it would hang.

Reply:
@ tedyu left comments on fixing this.